### PR TITLE
RD-2191 Fetch less data in Deployments View

### PR DESCRIPTION
**Pull request by Gelio**
This PR changes the parameters when fetching data in the Deployments View widget to only include a specific set of fields in the responses.

This reduces the amount of data received by the UI, since the `workflows`, `policy_triggers`, and other fields are not fetched. Take a look at the diff in `various-statuses.json` to see the difference - it's quite substantial :smile: 

Unfortunately, using some fields with `_include` results in a 500 backend error. I have reported it in RD-2900. Until that is fixed, this branch will not work.

## System tests

I will run them after RD-2900 is fixed since they will be failing at the moment.

### CODE REVIEW

File: widgets/common/src/deploymentsView/detailsPane/DrilldownButtons.tsx [hunk 1]
```diff
@@ -13,8 +13,21 @@ import {
 import { SubdeploymentStatusIcon } from '../StatusIcon';
 import type { Deployment, DeploymentStatus } from '../types';
 
+// NOTE: use a constrained identity function to make sure the array members match the properties
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/detailsPane/DrilldownButtons.tsx:16]: There will be conflicts with #1528 since that PR removes this file by splitting it into several smaller files

File: widgets/common/src/deploymentsView/types.ts [hunk 2]
```diff
@@ -25,7 +25,6 @@ export interface Deployment {
     blueprint_id: string;
     latest_execution_status: LatestExecutionStatus;
     deployment_status: DeploymentStatus;
-    installation_status: InstallationStatus;
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/types.ts:28]: Turns out this field was not used anywhere in the UI

File: widgets/common/src/deploymentsView/detailsPane/DrilldownButtons.tsx [hunk 3]
```diff
@@ -13,8 +13,21 @@ import {
...
+// of the `DrilldownButtonDeployment` type
+const drilldownButtonDeploymentKeys = (<T extends (keyof Deployment)[]>(keys: T) => keys)([
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/detailsPane/DrilldownButtons.tsx:18]: The constrained identity immediately invoked function expression (that's a mouthful :smile: CIF + IIFE) is there to make sure that `DrilldownButtonDeployment` contains only the properties specified in this array, and this array is used in the `_include` param.

Thus, a full type safety guarantee :slightly_smiling_face:

File: widgets/common/src/deploymentsView/DeploymentsView.tsx [hunk 4]
```diff
@@ -44,6 +45,29 @@ const reactSplitPaneOffset = 30;
...
+    'sub_services_status',
+    'sub_environments_count',
+    'sub_environments_status'
+]);
+
+type MustBeTrue<T extends true> = T;
+// @ts-expect-error Only for type-checking
+type AllPropertiesIncluded = MustBeTrue<
+    keyof Deployment extends typeof deploymentPropertiesToInclude[number] ? true : false
+>;
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/DeploymentsView.tsx:69]: I have tentatively added this piece of code to make sure that all the properties of `Deployment` are present in `deploymentPropertiesToInclude`.

If some property will be omitted, there will be a TS error, e.g. after removing `id`:

image

This check can be removed if you decide it's too complex :smile: I just added it as I wanted to experiment with TS a bit

**Comment by Gelio** [comment]
RD-2900 seems to be fixed, so I'll mark this PR as ready for review. Keep in mind that I have not verified it on a Manager that has RD-2900 fixed, since there weren't any builds with the fixes, so it still needs to be verified. As long as the system tests pass, it should be good to merge :slightly_smiling_face:

**Review by qooban — APPROVED** [review]
Nicely done!

Looking at the lines removed from `various-statuses.json` it's a big decrease in the size of data transferred. Great!

**Comment by Gelio** [comment]
Thanks for a review :smile: 

I managed to resolve the conflicts after merging #1528 - and I managed to fetch even less data (not refetch `id` and `display_name` when fetching subdeployment info). After the refactor done in #1528 it looked like some of the names did not make sense, so I renamed them.

I hope it's not too hard to review :slightly_smiling_face: Now the changes are more closely scoped (the changes are done inside of a module and are "private" to other modules)

Anyway, I'll leave the PR to you guys, I hope it will be straightforward to merge it (test on a manager that has RD-2900 fixed, run system tests, and then merge, hopefully) :smile:

**Comment by qooban** [comment]
Triggered system tests to verify the solution with the latest Cloudify Manager: [Stage-UI-System-Test #1017](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/1017/).

**Comment by qooban** [comment]
I discovered two problems:
1. Deployment URL was not constructed properly (it was "/deployments/nodecellar?all_sub_deployments=false?_include=sub_environments_count&_include=sub_environments_status&_include=sub_services_count&_include=sub_services_status" - question mark used twice + multiple `_include` paramters) - fixed it https://github.com/cloudify-cosmo/cloudify-stage/pull/1529/commits/241adb7269914dbec91dcd2570765b44c4aa2a9a
2. None of the following: 'sub_environments_count', 'sub_environments_status', 'sub_services_count', 'sub_services_status' are working with `_include` parameter, so that's still missing part on the manager side.

So I'm putting this PR back to draft until we've got resolution on the manager side.

**Comment by qooban** [comment]
@tehasdf Could you elaborate on p. 2 - https://github.com/cloudify-cosmo/cloudify-stage/pull/1529#issuecomment-890779094 ?

File: widgets/common/src/deploymentsView/detailsPane/drilldownButtons/subdeployments-result.ts [hunk 5]
```diff
@@ -12,13 +12,18 @@ const subdeploymentInfoDeploymentKeys = (<T extends (keyof Deployment)[]>(keys:
 
 type SubdeploymentInfo = Pick<Deployment, typeof subdeploymentInfoDeploymentKeys[number]>;
 
-const getDeploymentUrl = (id: string) => `/deployments/${id}?all_sub_deployments=false`;
+const getDeploymentUrl = (id: string) => `/deployments/${id}`;
```
Comment (Gelio) [inline widgets/common/src/deploymentsView/detailsPane/drilldownButtons/subdeployments-result.ts:15]: Welp, rookie mistake on my part :smile:

**Comment by qooban** [comment]
After discussing the problem with Lukasz, he suggested that issue from p. 2 can be solved by just adding `id` value to `_include` set in `GET /deployments/<id>` request. Change on the manager side would be harder to implement. 

Updated the code and triggered system tests to verify the solution with the latest Cloudify Manager: [Stage-UI-System-Test #1018](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/1018).

**Review by stepek — APPROVED** [review]
code LGTM, but tests failed

**Review by kubama — APPROVED** [review]

**Comment by qooban** [comment]
Thanks for the reviews.

> code LGTM, but tests failed

You're right. I'll check it out tomorrow and provide feedback on what is the reason of the failure.

**Comment by qooban** [comment]
I've managed to reproduce the problem in my local environment. Seems like a bug in the manager, details provided in RD-2923.

As for now I'm not merging this PR. I'll wait for the answer from the backend team.

**Comment by qooban** [comment]
As RD-2923 was fixed on the backend side, I triggered system tests to check: [Stage-UI-System-Test #1035](https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1035/pipeline).

If they will pass, I'm going to merge this PR.

**Comment by qooban** [comment]
TL;DR:
As in the second run of system tests [Stage-UI-System-Test #1038](https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1038/pipeline) (triggered today) all tests passed, I'm merging this PR, despite that in [Stage-UI-System-Test #1036](https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1036/pipeline) (triggered yesterday) 2 tests failed.

More details:
In [Stage-UI-System-Test #1036](https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1036/pipeline) which I ran yesterday two tests failed - Deployments widget and Topology widget specs. I triggered one more build and in [Stage-UI-System-Test #1038](https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1038/pipeline) all passed. I did a quick look at the Build-Images pipeline and it seems that there was some traffic yesterday afternoon, new images were created, so I suspect either there was a bug in the manager or those two failures were just one time issue.
